src/biogeochem/Fan2CTSMMod.F90

```diff
@@ -0,0 +1,1059 @@
+    ! timestep-cumulative (gN/m2) aging flux out of the oldest class
+    real(r8) :: n_residual, n_residual_total
+    ! H ion concentrations for grz and slr pools (== prescribed + soil pH for oldest class)
+    real(r8) :: Hconc_grz(num_cls_grz), Hconc_slr(num_cls_slr)
+    real(r8) :: fert_inc_tan, pH_soil, pH_crop
+
+    logical :: do_balance_checks
+    real(r8) :: tg, theta, thetasat, infiltr_m_s, evap_m_s, runoff_m_s, org_n_tot, &
+         nstored_old, nsoilman_old, nsoilfert_old, fert_to_air, fert_to_soil, fert_total, fert_urea, fert_tan, &
+         soilflux_org, urea_resid
+    ! TAN production flux from urea. Include one extra flux for the residual flux out of U2.
+    real(r8) :: tanprod_from_urea(num_cls_urea + 1)
+    real(r8) :: ureapools(num_cls_urea)
+    real(r8) :: fract_urea, fract_no3, soilph_min, soilph_max, &
+         soilpsi, fert_no3, fert_generic, bsw
+    integer :: def_ph_count
+    
+    Hconc_grz(1:num_cls_grz-1) = Hconc_grz_def
+    Hconc_slr(1:num_cls_slr-1) = Hconc_slr_def
+
+    soilph_min = 999
+    soilph_max = -999
+    def_ph_count = 0
+    dt = real(get_step_size(), r8)
+    do_balance_checks = balance_check_freq > 0 .and. mod(get_nstep(), balance_check_freq) == 0
+
+    associate(&
+         ngrz => soilbiogeochem_nitrogenflux_inst%manure_n_grz_col, &
+         man_u_grz => soilbiogeochem_nitrogenstate_inst%manure_u_grz_col, &
+         man_a_grz => soilbiogeochem_nitrogenstate_inst%manure_a_grz_col, &
+         man_r_grz => soilbiogeochem_nitrogenstate_inst%manure_r_grz_col, &
+         man_u_app => soilbiogeochem_nitrogenstate_inst%manure_u_app_col, &
+         man_a_app => soilbiogeochem_nitrogenstate_inst%manure_a_app_col, &
+         man_r_app => soilbiogeochem_nitrogenstate_inst%manure_r_app_col, &
+         ns => soilbiogeochem_nitrogenstate_inst, &
+         nf => soilbiogeochem_nitrogenflux_inst, &
+         cnv_nf => cnveg_nitrogenflux_inst, &
+         ram1 => frictionvel_inst%ram1_patch, &
+         rb1 => frictionvel_inst%rb1_patch)
+
+    nf%fert_n_appl_col(bounds%begc:bounds%endc) = 0.0
+    nf%manure_n_appl_col(bounds%begc:bounds%endc) = 0.0
+    nf%manure_tan_appl_col(bounds%begc:bounds%endc) = 0.0
+
+    call p2c(bounds, num_soilc, filter_soilc, &
+         cnv_nf%synthfert_patch(bounds%begp:bounds%endp), &
+         nf%fert_n_appl_col(bounds%begc:bounds%endc))
+
+    nf%manure_n_appl_col(bounds%begc:bounds%endc) = 0.0_r8
+    
+    if (do_balance_checks) then
+       nstored_old = get_total_n(ns, nf, 'pools_storage')
+       nsoilman_old = get_total_n(ns, nf, 'pools_manure')
+       nsoilfert_old = get_total_n(ns, nf, 'pools_fertilizer')
+    end if
+
+    ! Assign the "pastoral" manure entirely to the natural vegetation column
+    do fc = 1, num_soilc
+       c = filter_soilc(fc)
+       l = col%landunit(c)
+       if (.not. (lun%itype(l) == istsoil .or. lun%itype(l) == istcrop)) cycle
+       if (.not. col%active(c) .or. col%wtgcell(c) < 1e-6) cycle
```

Comment:
Currently some of the inputs are defined in mass per gridcell area, and they need to be converted to mass per column area. As long as this remains the case, it is not possible to avoid checking the cell weight, or to have the zero-weight columns to behave "exactly same" as others.

Reply:
I think the right solution here – long-term – is to ensure that the inputs are specified in terms of mass per crop column area; then we shouldn't need these weight multipliers or checks in the code. This is important in CTSM for a variety of reasons, a big one being that FAN's input data set cannot know ahead of time how much crop area there may be in each grid cell, since that is specified independently. It's generally more important to have per-area application rates correct than it is to have total grid cell application volumes correct. Imagine, for example, that crop area doubled relative to present day. It sounds like, with the way things are set up now, that would mean that inputs would be halved on a per-area basis; this is typically not what we want.

For now, while FAN is an experimental feature, this can be left as is. But if we want FAN to become more widely supported or default, then I think we'll want to change this.

Reply:
There are two types of manure N in FAN: the N in "mixed" livestock systems, and the N in "pastoral"  livestock systems. The "mixed N" is now given per crop area, so the problem doesn't occur there, but "pastoral N" is per land area. The reason is that the pastoral systems are normally not associated with crops but rather with semi-natural rangelands, often in regions that are not very suitable for growing crops. It is non-trivial to me which sub-gridcell area they should be associated with: the total native vegetation column, or just some pfts in it? 

I think that scaling the manure N with land use in transient runs is somewhat questionable in any case; it would be better to have it changing with time the same way as fertilizers are. There are datasets for the historical period, and IAMs include proxies that can be used for the future. We had an AGU poster which included some initial simulations for the future scenarios.

Reply:
I think this warrants an in-person / phone / zoom discussion at some point. My impression, though, is that this doesn't need to be settled now, and we can revisit it later. For now I just wanted to flag this as an issue that we'll need to revisit.

Reply:
(For the record, it looks like this has not been resolved. As I note above, I'm okay with that for now.)